server/machine_interfaces/haas.py

```diff
@@ -0,0 +1,192 @@
+# Luke Minton, MDE Fall 2021, VT ISE Learning Factory, Interface from Haas CNC machine to LF server
+# Machine currently has a serial (RS232) to USB port
+
+import serial, time, sys, zmq, pickle
+from typing import Dict, Any
+from threading import Thread
+from base import MachineInterface
+from lf_utils import retry
+
+# Haas macros for pulling data
+COOLANT_LEVEL = b"?Q600 1094\r\n"
+SPINDLE_SPEED = b"?Q600 3027\r\n"
+MACHINE_COORD_X = b"?Q600 5021\r\n"
+MACHINE_COORD_Y = b"?Q600 5022\r\n"
+MACHINE_COORD_Z = b"?Q600 5023\r\n"
+MACHINE_COORD_A = b"?Q600 5024\r\n"
+MACHINE_COORD_B = b"?Q600 5025\r\n"
+WORK_COORD_X = b"?Q600 5041\r\n"
+WORK_COORD_Y = b"?Q600 5042\r\n"
+WORK_COORD_Z = b"?Q600 5043\r\n"
+WORK_COORD_A = b"?Q600 5044\r\n"
+WORK_COORD_B = b"?Q600 5045\r\n"
+
+# Inherit from the Machine Interface class defined in base.py
+class HaasInterface(MachineInterface):
+    
+    def __init__(self, publish_port: int, machine_config: Dict, serial_port: str):
+
+        # set machine config
+        self._machine_config = machine_config
+
+        # set serial port name (i.e. COM3)
+        self._serial_port = serial.Serial(port = serial_port, baudrate = 9600, bytesize = serial.SEVENBITS, xonxoff = True, timeout = 5)
+
+        # Attempt to open a serial port for communication with the Haas
+        try:
+            self._serial_port.open()
+        except serial.SerialException:
+            if self._serial_port.is_open:
+                try:
+                    print("Port was open, attempting to close and reopen")
+                    self._serial_port.close()
+                    time.sleep(2)
+                    self._serial_port.open()
+                except:
+                    print("Failed to close port")
+                    event.clear()
+            else:
+                print("Failed to connect to serial port, either in use or does not exist")
+                event.clear() 
+
+        # set machine name
+        self._machine_name = self._machine_config["Haas CNC"]
+
+        # set up publish address
+        self._publish_address = f"tcp://127.0.0.1:{publish_port}"
+
+        # set up publish socket
+        self._context = zmq.Context()
+        self._publish_socket = self._context.socket(zmq.PUB)
+        retry(
+            self._publish_socket.connect,
+            self._publish_address,
+            handled_exceptions=zmq.error.ZMQError
+        )
+
+        # initialize thread
+        Thread.__init__(self)
+        self._stopped = False
```

Comment:
this code is from the base class. you should not be copying it here. use inheritance to use the functionality of the base class by putting `super().__init__()` at the top of this init method.

this is an ok intro tutorial: https://www.w3schools.com/python/python_inheritance.asp